### DISCUSSION TITLE
jinja2-fragments

### DISCUSSION THREAD

**Discussion by anryangelov** [volfpeter/fasthx#52] [Q&A] [answered]
Very helpful library, I have maybe stupid questions, can this library works with [jinja2-fragments](https://github.com/sponsfreixes/jinja2-fragments). And how forms and forms input values are handled as well.

**Comment by volfpeter** [comment]
Hi,

Not a stupid question at all. I think it was already asked both here and on Reddit. `fasthx` uses `fastapi.templating.Jinja2Templates` (= `starlette.templating.Jinja2Templates`) to render templates. If you can get `jinja2-fragments` working with `Jinja2Templates`, then it should work with `fasthx` as well, but there is no official support for it. Some reasoning behind this:

- There are many ways to modularize Jinja2 templates, and you don't really need fragments support if you did modularization right. Modularization is good practice anyway, it helps with maintenance, consistency, etc.
- HTMX, by its nature, also encourages and enables modularization, not just for HTML snippets, but also for APIs. Usually you end up with relatively small UI components/templates, and no need for fragments-like features at all.
- If you still need a fragments-like feature, then you can instead use the `hx-select` attribute of HTMX, which basically achieves the same thing, but in the frontend (where it probably belongs).

I'm not sure what you mean by "how forms ... are handled".

Do you mind if I convert this issue into a discussion? That way maybe others will also be more encouraged to join the conversion and help you.

**Comment by anryangelov** [comment]
Hi [volfpeter](https://github.com/volfpeter) thank you for the quick response. I would like to follow this pattern [Template Fragments
](https://htmx.org/essays/template-fragments/) and also [Locality of Behaviour (LoB)
](https://htmx.org/essays/locality-of-behaviour/). This is from the author of the HTMX. Having a lot of small templates everyone in separate files which are used only in one place lead to quite a few small template fragments which, in total, become difficult to manage and to reason about. This is django implementation of that [Adding template fragments or partials for the DTL](https://forum.djangoproject.com/t/adding-template-fragments-or-partials-for-the-dtl/21500). Using hx-select is option but rendering whole page again it is not elegant for me and easy for debugging. I think It would be super useful if there is option for selecting block something like this `@jinja.hx(index.html#user-list)`

**Reply by volfpeter** [reply, marked as answer]
Notice that the usage example from your comment does not require any change to the `fasthx` API. So I'll refer back to my original response:

> ... `fasthx` uses `fastapi.templating.Jinja2Templates` (= `starlette.templating.Jinja2Templates`) to render templates. If you can get `jinja2-fragments` working with `Jinja2Templates`, then it should work with `fasthx` as well ...

If you also have a look at the [FastAPI integration](https://github.com/sponsfreixes/jinja2-fragments/blob/main/src/jinja2_fragments/fastapi.py) of `jinja2-segments`, you will immediately see that it's done by sub-classing `Jinja2Templates`, which means you can use `jinja2-segments` with `fasthx` out of the box, you just need to pass a `Jinja2Blocks` instance to `fasthx.Jinja`, instead of `fastapi.templating.Jinja2Templates`.
